python/gps/algorithm/algorithm_traj_opt.py

```diff
@@ -98,15 +105,107 @@ def _stepadjust(self, m):
+                    for t in xrange(self.T - 1):
+                        diff = traj.k[t, :] + \
+                                traj.K[t, :, :].dot(sample_i_X[j, t, :]) - sample_i_U[j, t, :]
+                        samples_prob[i][itr_i, t, j] = -0.5 * np.sum(diff * (traj.inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                                                        np.sum(np.log(np.diag(traj.chol_pol_covar[t, :, :])))
+
+            # Evaluate sample prob under demo distribution.
+            for itr_i in xrange(Md):
+                for j in range(sample_i_X.shape[0]):
+                    for t in xrange(self.T - 1):
+                        diff = demo_traj[itr_i].k[t, :] + \
+                                demo_traj[itr_i].K[t, :, :].dot(sample_i_X[j, t, :]) - sample_i_U[j, t, :]
+                        samples_prob[i][itr + itr_i, t, j] = -0.5 * np.sum(diff * (demo_traj[itr_i].inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                                                        np.sum(np.log(np.diag(demo_traj[itr_i].chol_pol_covar[t, :, :])))
+            # Sum over time.
+            samples_prob_T[i] = np.sum(samples_prob[i], 1)
+            # Sum over the distributions.
+            samples_sum[i] = logsum(samples_prob_T[i], 0)
+
+        # Assume only one condition for the samples.
+        assert Md == M or M == 1
+        for idx in xrange(Md):
+            if M == 1:
+                i = 1
+            else:
+                i = idx
+            # Evaluate demo prob. under sample distributions.
+            for itr_i in xrange(itr):
+                traj = self.traj_distr[itr_i][i]
+                for j in xrange(demoX[idx].shape[0]):
+                    for t in xrange(self.T - 1):
+                        diff = traj.k[t, :] + \
+                                traj.K[t, :, :].dot(demoX[idx][j, t, :]) - demoU[idx][j, t, :]
+                        demos_prob[idx][itr_i, t, j] = -0.5 * np.sum(diff * (traj.inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                                                        np.sum(np.log(np.diag(traj.chol_pol_covar[t, :, :])))
+            # Evaluate demo prob. under demo distributions.
+            for itr_i in xrange(Md):
+                for j in range(demoX[idx].shape[0]):
+                    for t in xrange(self.T - 1):
+                        diff = demo_traj[itr_i].k[t, :] + \
+                                demo_traj[itr_i].K[t, :, :].dot(demoX[idx][j, t, :]) - demoU[idx][j, t, :]
+                        demos_prob[idx][itr + itr_i, t, j] = -0.5 * np.sum(diff * (demo_traj[itr_i].inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                                                        np.sum(np.log(np.diag(demo_traj[itr_i].chol_pol_covar[t, :, :])))
+            # Sum over time.
+            demos_prob_T[idx] = np.sum(demos_prob[idx], 1)
+            # Sum over the distributions.
+            demos_sum[idx] = logsum(demos_prob_T[idx], 0)
+
+        self.samples_prob = samples_sum
```

Comment:
Looking at the code, they are only used later in the function, so let's not store them.

Also, let's change the naming to use the variable names demos_logprob and samples_logprob instead of demos_prob and samples_prob, and replace the new variables demos_prob_T and demos_sum with in-place computation, e.g. demos_logiw[idx] = logsum(np.sum(demos_logprob[idx], 1), 0), where iw stands for importance weight.